controllers/s3/s3bucket.go

```diff
@@ -0,0 +1,244 @@
+	// we want to let the resource.ManagedReconciler know so it can create it.
+	if s3.IsErrorNotFound(err) {
+		return resource.ExternalObservation{ResourceExists: false}, nil
+	} else if err != nil {
+		return resource.ExternalObservation{}, errors.Wrap(err, "cannot get instance")
+	}
+
+	// Update our "Ready" status condition to reflect the status of the external
+	// resource. Most managed resources use the below well known reasons that
+	// the "Ready" status may be true or false, but managed resource authors
+	// are welcome to define and use their own.
+	switch bucket.Status.AtProvider.Status {
+	case statusOnline:
+		// If the resource is available we also want to mark it as bindable to
+		// resource claims.
+		bucket.SetConditions(runtimev1alpha1.Available())
+		resource.SetBindable(bucket)
+	case statusCreating:
+		bucket.SetConditions(runtimev1alpha1.Creating())
+	case statusDeleting:
+		bucket.SetConditions(runtimev1alpha1.Deleting())
+	default:
+		bucket.Status.AtProvider.Status = statusCreating
+	}
+
+	accessKey, secretKey, err := s3.GetKeys(bucketUser)
+	if err != nil {
+		return resource.ExternalObservation{}, err
+	}
+	bucket.Status.AtProvider.ObjectUserID = bucketUser.ID
+	bucket.Status.AtProvider.Status = statusOnline
+
+	// Finally, we report what we know about the external resource. Any
+	// ConnectionDetails we return will be published to the managed resource's
+	// connection secret if it specified one.
+	o := resource.ExternalObservation{
+		ResourceExists:   true,
+		ResourceUpToDate: true,
+		ConnectionDetails: resource.ConnectionDetails{
+			runtimev1alpha1.ResourceCredentialsSecretUserKey:     []byte(accessKey),
+			runtimev1alpha1.ResourceCredentialsSecretPasswordKey: []byte(secretKey),
+			runtimev1alpha1.ResourceCredentialsSecretEndpointKey: []byte(fmt.Sprintf(s3.S3EndpointFormat, bucket.Spec.ForProvider.Region)),
+			resourceCredentialsSecretBucketname:                  []byte(bucketName),
+		},
+	}
+
+	return o, nil
+}
+
+// Create a new external resource based on the specification of our managed
+// resource. resource.ManagedReconciler only calls Create if Observe reported
+// that the external resource did not exist.
+func (e *external) Create(ctx context.Context, mg resource.Managed) (resource.ExternalCreation, error) {
+	bucket, ok := mg.(*storagev1alpha1.S3Bucket)
+	if !ok {
+		return resource.ExternalCreation{}, errors.New(errNotInstance)
+	}
+	log.Info("Create", "bucket", bucket.Name)
+
+	bucketName := meta.GetExternalName(bucket)
+	objectUser, err := e.s3Client.CreateOrUpdateBucket(ctx, bucket.Status.AtProvider.ObjectUserID, bucketName, bucket.Spec.ForProvider.Region, bucket.Spec.ForProvider.CannedACL, bucket.Spec.ForProvider.Tags)
+	if err != nil {
+		return resource.ExternalCreation{}, errors.Wrap(err, "cannot create instance")
+	}
+
+	bucket.Status.AtProvider.ObjectUserID = objectUser.ID
+	bucket.Status.AtProvider.Status = statusOnline
```

Comment:
Typically we reserve `AtProvider` for status observations that come from the actual provider (i.e. that are pretty much direct translations of the cloud provider API). This status seems to be a concept introduced by this controller - perhaps it should be directly under `bucket.Status`?